### PR TITLE
Linear-chain CRF

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented crf loss function which calculates negative log-likelihood of linear-chain CRF. It also contains Viterbi algorithm to find the best path.
And this PR contains log-sum-exp function and argmin/argmax function.

### CODE REVIEW

**Comment by unnonouno** [comment]
I split this PR into #1216 and #1217

**Comment by delta2323** [comment]
Should we review #1216 and #1217 first?

**Comment by unnonouno** [comment]
Yes. I'll rebase this pr after that.

**Comment by rezoo** [comment]
Thank you for PR! Although I do not review the code w.r.t. argmax/argmin functions (I would like to pass this work to #1217), it seems that this code works well.

A trivial problem I have felt through reading this PR is the use of the name of "viterbi".
It is known that Viterbi algorithm is used to obtain the state that maximizes a density of CRF model, however,
I consider that it is better to replace this name with more definite ones such as "argmax" and "map_inference".
(It would be also nice to add a description that argmax_crf1d solves the maximization problem with the Viterbi algorithm to this function).
What do you think about this opinion?

File: tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+import numpy
+
+import chainer
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+
+
+class TestCRF1d(unittest.TestCase):
```
Comment (rezoo) [inline tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py]: How about adding new tests regarding a toy-example that can analytically compute the maximum of the CRF model and its state?

File: chainer/functions/loss/crf1d.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+def crf1d(cost, xs, ys):
+
+    """Calculates negative log-likelihood of linear-chain CRF.
+
+    It takes a transition cost matrix, a sequence of costs and a sequence of
+    labels. Let :math:`c_{st}` is transition cost from a label :math:`s` to
+    a label :math:`t`, :math:`x_{it}` is a cost value of a label :math:`t` at
+    position :math:`i`, and :math:`y_i` is an expected label value at position
+    :math:`i`. The negative log-likelihood of linear-chain CRF is defined as
```
Comment (rezoo) [inline chainer/functions/loss/crf1d.py]: It takes a transition cost matrix, a sequence of costs, and a sequence of labels.
Let :math:`c_{st}` be a transition cost from a label :math:`s` to a label :math:`t`,
:math:`x_{it}` be a cost of a label :math:`t` at position :math:`i`, and
:math:`y_i` be an expected label at position :math:`i`.

File: chainer/functions/loss/crf1d.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    .. math::
+        L = -\\left( \\sum_{i=1}^l x_{iy_i} + \\
+             \\sum_{i=1}^{l-1} c_{y_i y_{i+1}} - {\\log(Z)} \\right) ,
+
+    where :math:`l` is the length of the input sequence and :math:`Z` is the
+    partition function.
```
Comment (rezoo) [inline chainer/functions/loss/crf1d.py]: where :math:`l` is the length of the input sequence and :math:`Z` is the normalizing constant called the partition function.

File: chainer/functions/loss/crf1d.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    for y1, y2 in zip(ys[:-1], ys[1:]):
+        score += reshape.reshape(
+            embed_id.embed_id(y1 * n_label + y2, cost), (n_batch,))
+    for x, y in zip(xs, ys):
+        score += select_item.select_item(x, y)
+
+    return _sum.sum(logz - score) / n_batch
+
+
+def crf1d_viterbi(cost, xs):
```
Comment (rezoo) [inline chainer/functions/loss/crf1d.py]: I think it would be better to use ``argmax_crf1d`` instead of ``crf1d_viterbi``.

File: chainer/links/loss/crf1d.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    def __init__(self, n_label):
+        super(CRF1d, self).__init__(cost=(n_label, n_label))
+        self.cost.data[...] = 0
+
+    def __call__(self, xs, ys):
+        return crf1d.crf1d(self.cost, xs, ys)
+
+    def viterbi(self, xs):
```
Comment (rezoo) [inline chainer/links/loss/crf1d.py]: How about using ``argmax`` instead of ``viterbi``?

File: chainer/functions/loss/crf1d.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+            :class:`~chainer.Variable` holds a :math:`B \\times K`
+            matrix, where :math:`B` is mini-batch size, :math:`K` is the number
+            of labels.
+        ys (list of Variable): Expected output labels. Each
+            :class:`~chainer.Variable` holds a :math:`B` integer vector.
+
+    Returns:
+        ~chainer.Variable: A variable holding a :math:`B` float vector. Each
+            element stores negative log-likelihood of the corresponding
+            sequence.
```
Comment (rezoo) [inline chainer/functions/loss/crf1d.py]: It seems that the returned variable holds a scalar value.

File: tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py (hunk 1 above)
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py]: I checked all `3^3` states manually in `test_forward` and `test_viterbi`.

File: chainer/functions/loss/crf1d.py (hunk 6 above)
Comment (unnonouno) [inline chainer/functions/loss/crf1d.py]: Oh, that's true.

**Comment by unnonouno** [comment]
That's a good suggestion. I feel `argmax_crf1d` is better. I'll fix the name.

**Comment by delta2323** [comment]
Is CRF so common that we do not have to cite any paper. (For example, `F.BatchNormalization` cites the original paper in its document. Does CRF have the corresponding paper?)

**Comment by unnonouno** [comment]
The documentation is not enough yet. Please check interface and specification first.

File: tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py (hunk 1 above)
Comment (rezoo) [inline tests/chainer_tests/functions_tests/loss_tests/test_crf1d.py]: I got it.

**Comment by rezoo** [comment]
> Please check interface and specification first.

In terms of interface, I felt that it is more natural to change an input list ``[xs]`` with ``N`` variables to one variable that takes ``N`` channels.
However, I also considered that it is a relatively hard task; we have to implement ``inclusive_scan`` operator that can take any binary operators to ``cupy``, implement ``scan`` function using ``logsumexp`` operator, and compute marginal distributions used for ``backward()`` function.
Therefore, currently I am considering that it is better for us to employ this specification, and in near future we will support the both types of inputs.

**Comment by unnonouno** [comment]
I think we can support variable length inputs like #1209. I feel it is better for `crf1d` functions to support such inputs because users can simply combine RNNs and CRFs.

**Comment by rezoo** [comment]
I agree with you in the case of RNNs and CRFs. At least this function would be useful when we want to improve the accuracy of RNNs with CRFs.
@delta2323 anyway, I think it is OK to merge this PR unless you consider there exist any crucial  problems.

**Comment by delta2323** [comment]
OK, I do not think we must cite some paper.

**Comment by rezoo** [comment]
Thank you! Then I would like to merge it.

**Comment by unnonouno** [comment]
Thank you! BTW, Lafferty's paper i cited is the first one mensioning linear-chain CRF. And, I think this is the first work combining linear-chain CRF with neural netoworks. http://arxiv.org/pdf/1508.01991.pdf

**Comment by rezoo** [comment]
In order to describe their work as a "first work", I think we should carefully investigate it because the combination of CRFs and NNs is a very common approach.

For example, in the field of NLP I found that the following paper uses CRFs and RNNs for sequence-tagging problems:
K. Yao, et. al., Recurrent conditional random field for language understanding, In ICASSP, 2014.

In the other areas (e.g., ML), several papers are proposed for improving accuracy with the both NNs and CRFs:
Neural conditional random fields. In NIPS, 2010.
Conditional neural fields. In NIPS, 2009.
Supervised learning of image restoration with convolutional networks. In ICCV, 2007.

**Comment by unnonouno** [comment]
Oh, thank you. It is Ok to add these references to the document.
